lib/core/test/unit/Cardano/Wallet/Primitive/AddressDiscovery/DelegationSpec.hs

```diff
@@ -0,0 +1,350 @@
+            usableKeys (apply (chainFromCmds cmds) s0)
+            === usableKeys (apply (chainFromCmds $ filter (not . isAdversarial) cmds) s0)
+    it "chainFromCmds (cmdsFromChain (chainFromCmds cmds)) == chainFromCmds cmds"
+        $ property $ \cmds -> do
+            chainFromCmds (cmdsFromChain (chainFromCmds cmds))
+                === chainFromCmds (filter (not . isAdversarial) cmds)
+    it "activeKeys (apply (cmds <> CmdSetPortfolioOf 0) s0) === activeKeys s0"
+        $ property $ \cmds -> do
+            let chain = chainFromCmds (cmds ++ [CmdSetPortfolioOf 0])
+            let s = apply chain s0
+            activeKeys s === activeKeys s0
+    it "usableKeys (apply (cmds <> CmdSetPortfolioOf 0) s0) === usableKeys s0"
+        $ property $ \cmds -> do
+            let chain = chainFromCmds (cmds ++ [CmdSetPortfolioOf 0])
+            let s = apply chain s0
+            counterexample ("pointer state: " <> show (pointerState s))$
+                usableKeys s === usableKeys s0
+    it "presentableKeys (apply (cmds <> CmdSetPortfolioOf 0) s0) ⊇ presentableKeys s0"
+        $ property $ \cmds -> do
+            let chain = chainFromCmds (cmds ++ [CmdSetPortfolioOf 0])
+            let s = apply chain s0
+            let presentableKeys' = Set.fromList . presentableKeys
+            if any isAdversarial cmds
+            then label "with adversarial commands (⊇)" $
+                (presentableKeys' s0) `Set.isSubsetOf` (presentableKeys' s)
+                -- TODO: If presentable keys aren't influenced by key
+                -- registrations... then we could use === here too.
+            else label "no adversarial commands (=)" $
+                (presentableKeys' s0) === (presentableKeys' s)
+
+    describe "Mock ledger" $ do
+        it "accepts all generated mock chains" $ property $ \cmds -> do
+            let chain = chainFromCmds cmds
+            let l = applyLedger chain initialLedger
+            case l of
+                Right _ -> label "chain valid according to ledger" True
+                Left e -> counterexample e False
+
+        describe "Dropped or re-ordered transactions" $ do
+            it "chain valid to ledger => activeKeys == regs ledger" $ property $ \cmds -> do
+                let chain = chainFromCmds cmds
+                let showTxs txs =
+                        "\nAccepted txs:\n" <> unlines (map show txs) <> "\n"
+                forAllShow (sublistOf chain) showTxs $ \subChain -> do
+                    let s = apply subChain s0
+                    case (applyLedger subChain initialLedger) of
+                        Right (Ledger regs) -> counterexample
+                            "valid chain => expecting the wallet's activeKeys \
+                            \to match all registered keys in the ledger" $
+                            Set.fromList (fmap toRewardAccount (activeKeys s))
+                                === regs
+                        Left _ -> label "invalid chain" True
```

Comment:
Can now show that rollbacks are a problem 🎉 
```
 1) Mock ledger, Dropped or re-ordered transactions, chain valid to ledger => activeKeys == regs ledger
       Falsified (after 13 tests and 6 shrinks):
         [CmdSetPortfolioOf 1,CmdSetPortfolioOf 2]

         Accepted txs:
         Tx [RegisterKey (RewardAccount "1"),Delegate (RewardAccount "1")]


         valid chain => expecting the wallet's activeKeys to match all registered keys in the ledger
         fromList [] /= fromList [RewardAccount "1"]
```

Reply:
Seems `sublistOf` cannot re-order elements. Not sure if this matters. Maybe 🤔 Should either fix or add a comment at least.